API/COMPAT: add pydatetime-style positional args to Timestamp constructor

thejohnfreeman opened the pull request:
- [X] closes #10758
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

If more tests are desired, please let me know where to put them. Same for the whatsnew entry. This is my first pull request for Pandas.

cc @jreback

thejohnfreeman committed:
API/COMPAT: add pydatetime-style positional args to Timestamp constructor

jreback commented on pandas/tslib.pyx:
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
         cdef _TSObject ts
         cdef _Timestamp ts_base
 
+        if offset is not None and type(offset) is int:
```
use ``is_integer_object(offset)``

jreback commented on pandas/tslib.pyx:
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
-    def __new__(cls, object ts_input, object offset=None, tz=None, unit=None):
+    def __new__(cls,
+            # Keep legacy parameter names.
+            object ts_input, object offset=None, tz=None, unit=None,
```
put the comments below the actual constructor

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
             Timestamp(Period('1000-01-01'))
 
+    def test_constructor_positional(self):
+        with tm.assertRaises(TypeError):
```
add the issue number here

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
+    def test_constructor_positional(self):
+        with tm.assertRaises(TypeError):
+            Timestamp(2000, 1)
+        with tm.assertRaises(ValueError):
```
not sure if its worth it to try to make ``Timestamp(2000, 1)`` raise a ``ValueError``

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
+            Timestamp(2000, 1, 0)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 32)
+
```
add some tests where ``None`` is passed for various places ``Timestamp(2000, 1, 1, None, None)``
``Timestamp(2000, 1, None, ...)``

jreback commented on pandas/tslib.pyx:
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
+            # Keep legacy parameter names.
+            object ts_input, object offset=None, tz=None, unit=None,
+            # Switch to datetime positional paramers, starting with `minute`.
+            minute=0, second=0, microsecond=0, tzinfo=None):
```
These should default to ``None`` rather than ``0``, I *think* that ``datetime.datetime`` will handle that

thejohnfreeman replied:
`datetime.datetime` raises `TypeError` when `None` is passed for any argument (except `tzinfo`), which led me to default to `0`. Is that ok?

```
----> 1 dt.datetime(2012, 1, 1, None)

TypeError: an integer is required (got type NoneType)
```

jreback replied:
ok.

I would ideally I think like to have this signature:

``ts_input=None, offset=None, tz=None, unit=None, year=None, month=None, day=None, hour=None, minute=None, second=None, tzinfo=None``

then its simple to do kwargs, e.g. you MUST provide ``year`` is not None (and all of ts_input, offset, tz, unit must be ``None``).

Then if you are actually creating a ``datetime.time`` (thru either positional or kwargs), then you can just call like (after you have done a rename of args for the positional if needed).

e.g.

```datetime.datetime(year=year, month=month, day=day, hour=hour or 0, minute=minute or 0, second=second or 0)``

jreback commented:
pls add a whatsnew entry.

any luck with accepting the kwargs as well? (or too complicated)

thejohnfreeman commented:
Should the whatsnew go in for 18.1 or something else?

kwargs got too complicated. I wanted to reach a working version of positional parameters first before building kwargs on top of it. I will try out an idea this weekend.

jreback commented:
you can put in 0.18.1; if you finish up before we release we can just move

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -22,6 +22,14 @@ New features
 ~~~~~~~~~~~~
 
+- Timestamps accept positional parameters like `datetime.datetime`:
+
```
put ``Timestamp`` in double back-ticks 

say: can now accept

put a link in for ``datetime.datetime`` back to the constructor (on the python documentation page)

@jorisvandenbossche do we have a nicer way of doing this (as opposed to a straight link), IIRC, sphinx allows us to use an intra-module link (like we do with numpy)

jorisvandenbossche replied:
yes, you can just do ```:func:`datetime.datetime` ``` and that will link to the standard library's method

jreback commented:
@thejohnfreeman can you rebase / update & move whatsnew to 0.18.2

thejohnfreeman commented:
Ok, I rebased. I have time this week to finish this patch. I wrote a longer comment asking a question, but I'll just go ahead and try the signature you gave above.

```
def __new__(cls,
        object ts_input, object offset=None, tz=None, unit=None,
        year=None, month=None, day=None,
        hour=None, minute=None, second=None, tzinfo=None):
```

jreback commented on pandas/tslib.pyx:
```diff
@@ -288,10 +289,36 @@ class Timestamp(_Timestamp):
+        # The parameter list folds together legacy parameter names (the first
+        # four) and positional parameter names from pydatetime (starting with
+        # `minute`).
+        #
```
good, this is a nice way to do this.

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,36 @@ def test_constructor_invalid(self):
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
+
+        pos_args = [2000, 1, 2, 3, 4, 5, 999999]
```
can u also add another with ``tzinfo`` (another set of pos_args).

and another with kwargs as well.

jorisvandenbossche commented:
@thejohnfreeman You will also need to update the Timestamp docstring

jreback commented:
also I believe its trivial to close #11630 (or it might already be done). Add a test and should be straightforward.

thejohnfreeman committed:
Support positional and keyword arguments for Timestamp

thejohnfreeman commented:
Take a look at my attempt at supporting both positional and keyword arguments. I removed some tests for TypeError; it just means we are more permissive. I'll fix up the doc string and add the other requested tests tomorrow.

jreback commented on pandas/tslib.pyx:
```diff
@@ -288,10 +288,43 @@ class Timestamp(_Timestamp):
+            return Timestamp(datetime(year, month, day, hour or 0,
+                minute or 0, second or 0, microsecond or 0, tzinfo),
+                tz=tzinfo)
+        if is_integer_object(offset):
```
elif

jreback commented:
whoosh @thejohnfreeman nice constuctor now! yeah prob need some more tests.

shoyer commented:
Effectively, we are doing type dependent multiple dispatch here with the `Timestamp` constructor. To be honest, I'm not a big fan of adding multiple paths like this that do such different things with the arguments.

jreback commented:
@shoyer that doesn't make any sense. ``Timestamp`` is a sub-class of ``datetime``. It doesn't take the correct (positional, nor keyword args ATM). I don't see the problem here. Yes it also does a lot more, but that's exactly the point, it IS user friendly w/o be ambiguous.

thejohnfreeman committed:
Fix docstring; add tests for kwargs

thejohnfreeman commented:
I updated the docstring and added tests. I'm considering this a first draft submitted for approval. If it's approved, I'll squash the commits and we can wait for Travis.

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
+        actual = ts.to_pydatetime()
+        expected = datetime.datetime(2000, 1, 2)
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
```
what is the purpose of this test here? ``.to_pydatetime()`` is already tested elsewhere.

thejohnfreeman replied:
To test for equivalence. Let me know if I should use a different pattern.

jreback replied:
I understand but that's not part of this PR (and its already tested), so not necessary.

jreback commented on pandas/tslib.pyx:
```diff
@@ -225,8 +225,12 @@ class Timestamp(_Timestamp):
+    first, legacy form accepts four parameters. They can be passed by
+    position or keyword.
+
+    Legacy Parameters
```
don't call these ``Legacy`` Parameters; these are in fact the normal calling convention. The ``datetime.datetime`` convention are the compat parameters (but you don't have to call it differently).

jreback commented on pandas/tslib.pyx:
```diff
@@ -235,6 +239,20 @@ class Timestamp(_Timestamp):
+    be passed by either position or keyword, but not both mixed together.
+
+    datetime.datetime Parameters
+    ------------------------------
```
put a refernce to ``datetime.datetime`` doc page

thejohnfreeman replied:
What's the format for that?

jreback replied:
```
    .. versionadded:: 0.18.2
```

be sure to put a blank line before and after (otherwise we get warnings)

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
+    def test_constructor_positional(self):
+        # GH 10758
+        with tm.assertRaises(TypeError):
+            Timestamp(2000, 1)
```
can you use ``assertRaisesRegexp`` here to check for things

```
 In [1]: datetime.datetime(2000,1)
TypeError: Required argument 'day' (pos 3) not found
```

doesn't have to be an exact checking, but need to make sure that it is raising the ``datetime.datetime`` constructor error message and not something else.

thejohnfreeman replied:
For the cases where not enough parameters are passed to `Timestamp`, the error will be `an integer is required (got NoneType)` instead of `Required argument not found`. Matching the `datetime` exception exactly requires 3 additional checks and calls:

```
if year is None:
    return datetime.datetime()
if month is None:
    return datetime.datetime(year)
if day is None:
    return datetime.datetime(year, month)
```

which seems ugly. Do we really want that?

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
+        ts = Timestamp(year=2000, month=1, day=2)
+
+        actual = ts.to_pydatetime()
+        expected = datetime.datetime(year=2000, month=1, day=2)
```
this also seems superfluous (the ``.to_pydatetime()`` check)

jreback commented:
did you add a test for #11630 as well (I *think* you covered it in code, just need the test)

thejohnfreeman commented:
For #11630, I'm considering the test in test_constructor_positional to be equivalent:

```
ts = Timestamp(2000, 1, 2)
```

I can add a test with the exact parameters from #11630 too. Regardless, I will add the issue in a comment.

jreback commented:
@thejohnfreeman yes add that exact test (e.g. 

```
self.assertEqual(repr(Timestamp(2015, 11, 12)),  repr(Timestamp('20151112')))
```
should be enough

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -31,6 +31,14 @@ Other enhancements
 - The ``.tz_localize()`` method of ``DatetimeIndex`` and ``Timestamp`` has gained the ``errors`` keyword, so you can potentially coerce nonexistent timestamps to ``NaT``. The default behaviour remains to raising a ``NonExistentTimeError`` (:issue:`13057`)
 
+- ``Timestamp``s can now accept positional parameters like :func:`datetime.datetime`:
+
```
add the issue numbers at the end here

jreback commented on pandas/tslib.pyx:
```diff
@@ -225,8 +225,12 @@ class Timestamp(_Timestamp):
-    ----------
+    There are essentially three calling conventions for the constructor. The
+    first, legacy form accepts four parameters. They can be passed by
+    position or keyword.
```
add a versionadded tag here (0.18.2)

shoyer commented:
> @shoyer that doesn't make any sense. Timestamp is a sub-class of datetime. It doesn't take the correct (positional, nor keyword args ATM). I don't see the problem here. Yes it also does a lot more, but that's exactly the point, it IS user friendly w/o be ambiguous.

You cannot add alternate signatures without adding ambiguity. For example, suppose a naive user tries to pass three string arguments instead of integers into the new constructor. This is invalid for both datetime and the existing Timestamp signature, but because these arguments are supplied as strings, it would still trigger the existing codepath.

Let's compare what they would see:
```
>>> pd.Timestamp('2000', '01', '01')
UnknownTimeZoneError: '01'

>>> datetime.datetime('2000', '01', '01')
TypeError: an integer is required
```
The error from datetime is *much* more informative.

The standard library `datetime.datetime` class does not accept a string to convert to avoid exactly these sorts of issues. Instead, you are forced to use alternate constructors such as `strptime`.

It is unambiguously poor API design to make functions that do two very different things depending on the exact form of the input.

I'm all for making `pd.Timestamp` more consistent with `datetime.datetime` if you're willing to deal with the hassle of a long deprecation cycle, but the right approach is not to overload the constructor to do something totally different. The right approach would be to switch the constructor entirely.

thejohnfreeman commented:
Personally, I can sympathize with @shoyer. In all languages, I prefer that constructors just directly initialize fields and leave any magic to ordinary functions. Especially in Python, you only get one constructor but you can make unlimited functions for every style of construction you want, and it becomes easier to create custom test objects.

That said, I think it's more important to fit into a project's existing style and API, and I can see an argument for fitting this functionality into the `Timestamp` constructor. For now, I'm going to follow Jeff's instructions on this since that's what I've been doing all along.

thejohnfreeman committed:
review fixes: versionadded, issue links, repr tests, legacy

thejohnfreeman committed:
PEP8

shoyer commented:
What about deprecating passing a single object (string or datetime) to `pd.Timestamp` in favor of using `pd.to_datetime`? That way we could get back to only one constructor eventually.

It would not be the end of the world to add this alternate constructor but IMO it would not be a positive improvement to have both.

If to do do it, let's clearly document the two constructors at the top of the docstring for the class, e.g.,
```
Timestamp(ts_input, object_offset=None, tz=None, unit=None)
Timestamp(year, month, day[, hour[, minute[, second[, microsecond[, tzinfo]]]]])

Timestamp is...
```
I *think* sphinx can process that correctly but we should double check.

jreback commented:
@shoyer why would you blow up a very useful interface?

``Timestamp('20130101')`` is such a useful feature and you want to DEPRECATE?

the entire point is to REDUCE mental effort here. The whole notion of a user having to choose alternate constructors (and I know we had this discussion w.r.t. ``RangeIndex``) is completely bonkers to me.

Sure if things are ambiguous, then you have to choose. In this case it is wholly unambiguous, replicates the existing (python ``datetime.datetime`` API), and provides very useful parsing abilities, not to mention its very light weight.

shoyer commented:
> Timestamp('20130101') is such a useful feature and you want to DEPRECATE?

I agree -- my proposal is not entirely serious. I do think we need to make hard choices like this to achieve reliable APIs.

>the entire point is to REDUCE mental effort here. The whole notion of a user having to choose alternate constructors (and I know we had this discussion w.r.t. RangeIndex) is completely bonkers to me.

Just to prove I'm not crazy, here are some high lights from the first page of a Google search for "Python multiple constructors":
http://stackoverflow.com/questions/2164258/multiple-constructors-in-python
http://stackoverflow.com/questions/682504/what-is-a-clean-pythonic-way-to-have-multiple-constructors-in-python
http://www.erol.si/2016/01/python-and-multiple-constructors/
https://pythonconquerstheuniverse.wordpress.com/2010/03/17/multiple-constructors-in-a-python-class/ (read the comments)

I could go on, but you get the point.

You'll notice that the universal advice seems to be to skip the magic and use classmethods for alternate constructors instead of overloading `__init__`.

kawochen commented:
If we are to add more code paths in the constructor, is it possible to make the entry points more obvious by calling specific `from_*` methods inside `__new__` or `__init__`?  After reading the doc string I feel like it might become harder for a casual contributor like me to debug the code.  But I think this can be remedied by making the intention of each code path more visible in the code.

thejohnfreeman commented:
Hi @kawochen, I too want to make the code paths understandable. I had added some comments to each new path to document its intention. Did they help clarify at all?

thejohnfreeman commented:
I keep getting lint errors in the automated check. It doesn't say what the error is, and running trace8 locally doesn't give me any indication either.

```
The command "ci/script.sh" exited with 1.
1.00s$ ci/lint.sh
inside ci/lint.sh
discarding /Users/travis/miniconda/envs/pandas/bin from PATH
prepending /Users/travis/miniconda/envs/pandas/bin to PATH
NOT Linting
```

Any ideas what I need to do to get Travis to pass?

jreback commented:
it's not lint errors but a failed test

https://travis-ci.org/pydata/pandas/jobs/129782361

thejohnfreeman commented:
Ah, it expects `Timestamp(None)` to return Not a Time (`NaT`).

thejohnfreeman committed:
Special object for missing first argument

jreback commented on pandas/tslib.pyx:
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
+    year : int
+    month : int
+    day : int
+    hour : int [optional]
```
I think these need to be: ``int, optional`` (of this form)

jreback commented on pandas/tslib.pyx:
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
+    be passed by either position or keyword, but not both mixed together.
+
+    :func:`datetime.datetime` Parameters
+    ------------------------------
```
make this line extend the entire length (to end of text)

jreback commented on pandas/tslib.pyx:
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
     unit : string
         numpy unit used for conversion, if ts_input is int or float
+
+    The other two forms mimic the parameters from datetime.datetime. They can
```
put ``datetime.datetime`` in double-backticks

jreback commented on pandas/tslib.pyx:
```diff
@@ -288,10 +310,43 @@ class Timestamp(_Timestamp):
+        #
+        # There are three calling forms:
+        # - In the legacy form, the first parameter, ts_input, is required
+        # and may be datetime-like, str, int, or float. The second parameter,
```
can you format this like:

```
# - In the ......
#   next line here
#   next line here
....
```

jreback commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,50 @@ def test_constructor_invalid(self):
+        self.assertEqual(
+            repr(Timestamp(2015, 11, 12, 1, 2, 3, 999999)),
+            repr(Timestamp('2015-11-12 01:02:03.999999')))
+
```
can you add the assertion for ``Timestamp(None)`` as  well

thejohnfreeman committed:
Format comments, test for NaT

jreback commented:
lgtm. @thejohnfreeman nice explanations in the comments. I think it is clear.

@jorisvandenbossche @shoyer 

comments?

jreback commented:
@thejohnfreeman did you add the test for #11630 (e.g the ``repr`` expr from above)

thejohnfreeman commented on pandas/tseries/tests/test_tslib.py:
```diff
@@ -180,6 +180,52 @@ def test_constructor_invalid(self):
+        # GH 11630
+        self.assertEqual(
+            repr(Timestamp(2015, 11, 12)),
+            repr(Timestamp('20151112')))
```
Here is the test for #11630, Github didn't do the best job linking it to your earlier comment.

jreback commented:
thanks @thejohnfreeman great PR!

thejohnfreeman commented:
Thank you!
